sky/provision/cudo/instance.py

```diff
@@ -0,0 +1,222 @@
+                  config: common.ProvisionConfig) -> common.ProvisionRecord:
+    """Runs instances for the given cluster."""
+
+    pending_status = ['pend', 'init', 'prol', 'boot']
+
+    while True:
+        instances = _filter_instances(cluster_name_on_cloud, pending_status)
+        if not instances:
+            break
+        logger.info(f'Waiting for {len(instances)} instances to be ready.')
+        time.sleep(POLL_INTERVAL)
+    exist_instances = _filter_instances(cluster_name_on_cloud, ['runn'])
+    head_instance_id = _get_head_instance_id(exist_instances)
+
+    to_start_count = config.count - len(exist_instances)
+    if to_start_count < 0:
+        raise RuntimeError(
+            f'Cluster {cluster_name_on_cloud} already has '
+            f'{len(exist_instances)} nodes, but {config.count} are required.')
+    if to_start_count == 0:
+        if head_instance_id is None:
+            raise RuntimeError(
+                f'Cluster {cluster_name_on_cloud} has no head node.')
+        logger.info(f'Cluster {cluster_name_on_cloud} already has '
+                    f'{len(exist_instances)} nodes, no need to start more.')
+        return common.ProvisionRecord(provider_name='cudo',
+                                      cluster_name=cluster_name_on_cloud,
+                                      region=region,
+                                      zone=None,
+                                      head_instance_id=head_instance_id,
+                                      resumed_instance_ids=[],
+                                      created_instance_ids=[])
+
+    created_instance_ids = []
+    public_key = config.node_config['AuthorizedKey']
+
+    for _ in range(to_start_count):
+        instance_type = config.node_config['InstanceType']
+        spec = cudo_machine_type.get_spec_from_instance(instance_type, region)
+
+        node_type = 'head' if head_instance_id is None else 'worker'
+        try:
+            instance_id = cudo_wrapper.launch(
+                name=f'{cluster_name_on_cloud}-{node_type}',
+                ssh_key=public_key,
+                data_center_id=region,
+                machine_type=spec['machine_type'],
+                memory_gib=int(spec['mem_gb']),
+                vcpu_count=int(spec['vcpu_count']),
+                gpu_count=int(float(spec['gpu_count'])),
+                gpu_model=spec['gpu_model'],
+                tags={},
+                disk_size=config.node_config['DiskSize'])
+        except Exception as e:  # pylint: disable=broad-except
+            logger.warning(f'run_instances error: {e}')
+            raise
+        logger.info(f'Launched instance {instance_id}.')
+        created_instance_ids.append(instance_id)
+        if head_instance_id is None:
+            head_instance_id = instance_id
+
+    # Wait for instances to be ready.
+    retries = 12  # times 10 second
+    results = {}
+    for instance_id, _ in instances.items():
```

Comment:
This should be `created_instance_ids` instead?
```suggestion
    for instance_id in created_instance_ids:
```